docs/spec.md

```diff
@@ -140,240 +43,355 @@ stablehlo.func @main(
+  window_strides = dense<[2, 1]> : tensor<2xi64>,
+  padding = dense<[[0, 1], [0, 0]]> : tensor<2x2xi64>
+} : (tensor<4x2xi32>, tensor<2x2xi32>, tensor<i32>) -> tensor<4x2xi32>
 ```
 
-For example, for `replica_groups = [[0, 1], [2, 3]]` and `num_partitions = 2`,
-`cross_replica` will produce
-`[[(0, 0), (1, 0)], [(0, 1), (1, 1)], [(2, 0), (3, 0)], [(2, 1), (3, 1)]]`.
+### Constants
 
-#### cross_partition
-
-Only cross-partition communications happen within each process group. This
-strategy takes `partition_groups` - a list of lists of partition ids - and
-computes a Cartesian product of `partition_groups` by `replica_ids`.
-`partition_groups` must have unique elements and cover all `partition_ids`.
-More formally:
-
-```Python
-def cross_partition(partition_groups: List[List[PartitionId]]) -> List[List[ProcessId]]:
-  for partition_group in partition_groups:
-    for replica_id in replica_ids:
-      process_group = []
-      for partition_id in partition_group:
-        process_group.append((replica_id, partition_id))
-      yield process_group
+```ebnf
+Constant ::= BooleanConstant
+           | IntegerConstant
+           | FloatConstant
+           | ComplexConstant
+           | TensorConstant
+           | StringConstant
+           | EnumConstant
 ```
 
-For example, for `partition_groups = [[0, 1]]` and `num_replicas = 4`,
-`cross_partition` will produce
-`[[(0, 0), (0, 1)], [(1, 0), (1, 1)], [(2, 0), (2, 1)], [(3, 0), (3, 1)]]`.
+**StableHLO constants** have a literal and a type which together represent
+a StableHLO value. Generally, the type is part of the constant syntax, except
+when it's unambiguous (e.g. a boolean constant unambiguously has type `i1`,
+whereas an integer constant can have multiple possible types).
 
-#### cross_replica_and_partition
-
-Both cross-replica and cross-partition communications may happen within each
-process group. This strategy takes `replica_groups` - a list of lists of
-replica ids - and computes Cartesian products of each `replica_group` by
-`partition_ids`. `replica_groups` must have unique elements and cover all
-`replica_ids`. More formally:
+```ebnf
+BooleanConstant ::= BooleanLiteral
+BooleanLiteral  ::= 'true' | 'false'
+```
 
-```Python
-def cross_replica_and_partition(replica_groups: List[List[ReplicaId]]) -> List[List[ProcessId]]:
-  for replica_group in replica_groups:
-    process_group = []
-    for partition_id in partition_ids:
-      for replica_id in replica_group:
-        process_group.append((replica_id, partition_id))
-    yield process_group
+**Boolean constants** represent boolean values `true` and `false`. Boolean
+constants have type `i1`.
+
+```ebnf
+IntegerConstant   ::= IntegerLiteral ':' IntegerType
+IntegerLiteral    ::= ['-' | '+'] DecimalDigits
+                    | ['-' | '+'] '0x' HexadecimalDigits
+DecimalDigits     ::= decimalDigit {decimalDigit}
+HexadecimalDigits ::= hexadecimalDigit {hexadecimalDigit}
+decimalDigit      ::= '0' | ... | '9'
```

Comment:
What is the lowercase vs capital letters scheme used here? If it is that lowercase means it only produces terminals, should it be used for `booleanLiteral` above as well?